src/exact/ExactMapper.cpp

```diff
@@ -275,38 +275,45 @@ void ExactMapper::map(const Configuration& settings) {
             op->getParameter().at(0), op->getParameter().at(1),
             op->getParameter().at(2));
       } else {
-        const Edge cnot = {locations.at(static_cast<std::size_t>(gate.control)),
-                           locations.at(gate.target)};
+        const Edge controlledGate = {
+            locations.at(static_cast<std::size_t>(gate.control)),
+            locations.at(gate.target)};
 
-        if (architecture.getCouplingMap().find(cnot) ==
+        if (architecture.getCouplingMap().find(controlledGate) ==
             architecture.getCouplingMap().end()) {
-          const Edge reverse = {cnot.second, cnot.first};
+          const Edge reverse = {controlledGate.second, controlledGate.first};
           if (architecture.getCouplingMap().find(reverse) ==
               architecture.getCouplingMap().end()) {
-            throw QMAPException(
-                "Invalid CNOT: " + std::to_string(reverse.first) + "-" +
-                std::to_string(reverse.second));
+            throw QMAPException("Invalid controlled gate " + op->getName() +
+                                ": " + std::to_string(reverse.first) + "-" +
+                                std::to_string(reverse.second));
           }
           if (settings.verbose) {
-            std::cout
-                << i
-                << ": Added (direction-reversed) cnot with control and target: "
-                << cnot.first << " " << cnot.second << std::endl;
+            std::cout << i << ": Added (direction-reversed) controlled gate "
+                      << op->getName()
+                      << " with control and target: " << controlledGate.first
+                      << " " << controlledGate.second << std::endl;
           }
           qcMapped.h(reverse.first);
           qcMapped.h(reverse.second);
-          qcMapped.x(reverse.second,
-                     qc::Control{static_cast<qc::Qubit>(reverse.first)});
+          qcMapped.emplace_back<qc::StandardOperation>(
+              qcMapped.getNqubits(),
+              qc::Control{static_cast<qc::Qubit>(reverse.first)},
+              reverse.second, op->getType(), op->getParameter().at(0),
+              op->getParameter().at(1), op->getParameter().at(2));
           qcMapped.h(reverse.second);
           qcMapped.h(reverse.first);
```

Comment:
This is where it get's a bit tricky (the same actually holds for the heuristic mapper as well).
At the top of my head, I am not sure if this transformation of applying Hadamard gates to all inputs and outputs to reverse control and target holds for every possible controlled-gate.
Just as an example, where this might currently lead to suboptimal results: Any controlled-operation that has a diagonal unitary matrix, such as the controlled-Z, does not need Hadamard gates at all. You can simply flip control and target without changing the functionality.
As a first step (within this PR), I would be fine with that kind of overhead.

Could you verify whether the general transformation
```
controlled-U3(a,b,c) qc, qt
==
h qc; h qt;
controlled-U3(a,b,c) qt, qc;
h qc; h qt;
```
holds?